### PR TITLE
[eslint-plugin] feat: no-deprecated-components rule

### PR DESCRIPTION
**Pull request by styu**
#### Checklist

- [x] Includes tests
- [ ] Update documentation

#### Changes proposed in this pull request:

Adds an ESLint rule that flags particular deprecated blueprint components and recommends their v2 versions instead

#### Reviewers should focus on:

What other components am I missing?

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-deprecated-components.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import { TSESTree } from "@typescript-eslint/utils";
+
+import { createRule } from "./utils/createRule";
+
+const DEPRECATED_TO_NEW_MAPPING: { [deprecated: string]: string } = {
+    AbstractComponent: "AbstractComponent2",
+    AbstractPureComponent: "AbstractPureComponent2",
+    CollapsibleList: "OverflowList",
+    Popover: "Popover2",
+    Tooltip: "Tooltip2",
```
Comment (adidahiya) [inline packages/eslint-plugin/src/rules/no-deprecated-components.ts]:
```suggestion
    MultiSelect: "MultiSelect2",
    PanelStack: "PanelStack2",
    Popover: "Popover2",
    Select: "Select2",
    Suggest: "Suggest2",
    Tooltip: "Tooltip2",
```

File: packages/eslint-plugin/src/rules/no-deprecated-components.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                deprecatedImports.some(
+                    deprecatedImport => deprecatedImport.type === "namespace" && deprecatedImport.namespace === name,
+                ) && DEPRECATED_TO_NEW_MAPPING[property] != null
+            );
+        }
+
+        // Get the list of all deprecated imports from blueprint
+        return {
+            "Program > ImportDeclaration": (node: TSESTree.ImportDeclaration) => {
+                if (node.source.value !== "@blueprintjs/core") {
```
Comment (adidahiya) [inline packages/eslint-plugin/src/rules/no-deprecated-components.ts]: This needs to check for other packages as well, like `@blueprintjs/select`. We could maintain a list of packages which have deprecated exports at the top of this file.

File: packages/eslint-plugin/test/no-deprecated-components.test.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                },
+            ],
+        },
+        {
+            code: dedent`
+                import * as BP from "@blueprintjs/core";
+
+                export class MyClass extends BP.AbstractComponent {
+                }
+                class MyClass2 extends BP.AbstractComponent {}
```
Comment (adidahiya) [inline packages/eslint-plugin/test/no-deprecated-components.test.ts:78]: looks like two errors should be expected here instead of just one

**Review by adidahiya — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin/test/no-deprecated-components.test.ts (hunk 3 above)
Comment (styu) [reply packages/eslint-plugin/test/no-deprecated-components.test.ts:78]: hm.. it passes both if i pass 1 error or 2 errors below 🤔

File: packages/eslint-plugin/src/rules/no-deprecated-components.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    MultiSelect: "MultiSelect2",
+    PanelStack: "PanelStack2",
+    Popover: "Popover2",
+    Select: "Select2",
+    Suggest: "Suggest2",
+    Tooltip: "Tooltip2",
+};
+const PACKAGES_WITH_DEPRECATED_IMPORTS = ["@blueprintjs/core", "@blueprintjs/select"];
+
+type MessageIds = "nonDeprecated" | "deprecated";
```
Comment (adidahiya) [inline packages/eslint-plugin/src/rules/no-deprecated-components.ts]: I don't think we need two message IDs, if only one ("nonDeprecated") is used.

File: packages/eslint-plugin/src/rules/no-deprecated-components.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+// tslint:disable object-literal-sort-keys
+export const noDeprecatedComponentsRule = createRule<unknown[], MessageIds>({
+    name: "no-deprecated-components",
+    meta: {
+        type: "suggestion",
+        docs: {
+            description: "Recommends using non-deprecated versions of the component or constant instead",
```
Comment (adidahiya) [inline packages/eslint-plugin/src/rules/no-deprecated-components.ts]:
```suggestion
            description: "Reports on usage of deprecated Blueprint components and recommends migrating to their corresponding newer non-deprecated API alternatives.",
```

File: packages/eslint-plugin/src/rules/no-deprecated-components.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            requiresTypeChecking: false,
+            recommended: "error",
+        },
+        messages: {
+            nonDeprecated: "Replace deprecated component {{ deprecated }} with {{ nonDeprecated }}",
```
Comment (adidahiya) [inline packages/eslint-plugin/src/rules/no-deprecated-components.ts]:
```suggestion
            nonDeprecated: "Replace usage of deprecated component {{ deprecated }} with {{ nonDeprecated }}",
```

File: packages/eslint-plugin/src/rules/no-deprecated-components.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    Select: "Select2",
+    Suggest: "Suggest2",
+    Tooltip: "Tooltip2",
+};
+const PACKAGES_WITH_DEPRECATED_IMPORTS = ["@blueprintjs/core", "@blueprintjs/select"];
+
+type MessageIds = "nonDeprecated" | "deprecated";
+
+// tslint:disable object-literal-sort-keys
+export const noDeprecatedComponentsRule = createRule<unknown[], MessageIds>({
```
Comment (adidahiya) [inline packages/eslint-plugin/src/rules/no-deprecated-components.ts:31]: I think there should be some mention here (even if just for devs inspecting the rule source) of the fact that we're not actually reading the `@deprecated` JSDoc annotation, and thus the rule is quite distinct from the `deprecation/deprecation` ESLint rule.

**Review by adidahiya — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin/src/rules/no-deprecated-components.ts (hunk 4 above)
Comment (styu) [reply packages/eslint-plugin/src/rules/no-deprecated-components.ts]: ah yeah my original thought was if we had components that were removed with no replacement, but i didn't end up capturing those here

**Review by adidahiya — APPROVED** [review]

**Comment by adidahiya** [comment]
thanks @styu!

**Pull request merged**